test/components/form/show.spec.js

```diff
@@ -76,7 +76,9 @@ describe('FormShow', () => {
     testData.extendedForms.createPast(1, { xmlFormId: 'f', draft: true });
     testData.standardFormAttachments.createPast(1);
     return load('/projects/1/forms/f/draft/attachments')
-      .beforeEachResponse(app => {
+      .beforeEachResponse((app, { url }) => {
+        // TODO need to find a way to show loading in show.vue when dataset is loading
```

Comment:
@matthew-white I would need your help here

Reply:
Many components show their own loading message when they request an additional resource. For example, like `FormAttachmentList`, `PublicLinkList` is a child component of `FormShow`. `PublicLinkList` sends a request for `publicLinks`, which `FormShow` doesn't request, so it renders its own loading message just for `publicLinks`:

- `FormShow` requests `project`, `form`, `formDraft`, and `attachments`. Until it receives them, it shows a loading message.
- Even while `FormShow` is waiting for those resources, it renders the `<router-view>`. As a result, `PublicLinkList` is rendered right away. It sends its request for `publicLinks` and renders its loading message.
- Even though `PublicLinkList` is rendered, it won't be shown/visible until `FormShow` has received its resources.
- Then `PublicLinkList` will show its loading message until it has received `publicLinks`.
- If `PublicLinkList` receives `publicLinks` before `FormShow` has received its resources, then `PublicLinkList` will no longer show its loading message, but the user will continue to see the loading message in `FormShow`.

Unlike `PublicLinkList, `FormAttachmentList` historically hasn't requested any additional resource. However, now it will request `datasets`, so it will need to render a loading message. I would start by approaching it like this:

```diff
 <div v-show="attachments.dataExists" id="form-attachment-list" ref="dropZone">
   <!-- ^^^ We use v-show rather than v-if so that this.$refs.dropZone is in
   the DOM from the first render. -->
+  <loading :state="datasets.initiallyLoading"/>
+  <div v-show="datasets.dataExists">
     <div class="heading-with-button">
       <!-- ... -->
     </div>
     <table id="form-attachment-list-table" class="table">
       <!-- ... -->
     </table>
+  </div>
   <form-attachment-popups
```

(By the way, the `v-show` that's currently on `#form-attachment-list` is a historical artifact and could probably be removed. `FormShow` already specifies a `v-show` that will hide `FormAttachmentList` if `!attachments.dataExists`.)

I think it's important to think about loading messages, because there are Central users on slow connections. However, don't worry too much about testing them. Unless there's a fair amount of complexity involved, I don't usually test loading messages.

Reply:
To patch this test, feel free to navigate to a different page or to create different test data (for example, a published form rather than a draft one).

Reply:
I thought about this a little more, and I'm not sure it's essential that `FormAttachmentList` show a loading message during the request for `datasets`. `datasets` is only used to decide whether to render the "Restore Dataset Link" button, and I think it's fine if there's a delay between when the rest of the page appears and that button appears. In other words, I think we can show the page as soon as it's otherwise ready, then allow it to change slightly once `datasets` comes in (possibly showing the button). Just a thought — feel free to add the loading message if that seems better!

Reply:
Thanks for the help, I elected not to show loading message if getDatasets is still loading in FormAttachments component